packages/block-editor/src/components/list-view/block-rename-ui.js

```diff
@@ -0,0 +1,136 @@
+		const inputRef = useFocusOnMount();
+		const inputDescriptionId = useInstanceId(
+			ListViewBlockRenameUI,
+			`block-editor-list-view-block-node__input-description`
+		);
+
+		const formDescriptionId = useInstanceId(
+			ListViewBlockRenameUI,
+			`block-editor-list-view-block-node__form-description`
+		);
+
+		// Local state for value of input **pre-submission**.
+		const [ inputValue, setInputValue ] = useState( blockTitle );
+
+		const onKeyDownHandler = ( event ) => {
+			// Trap events to input when editing to avoid
+			// default list view key handing (e.g. arrow
+			// keys for navigation).
+			event.stopPropagation();
+
+			// Handle ENTER and ESC exits editing mode.
+			if ( event.keyCode === ENTER || event.keyCode === ESCAPE ) {
+				if ( event.keyCode === ESCAPE ) {
+					handleCancel();
+				}
+
+				if ( event.keyCode === ENTER ) {
+					handleSubmit();
+				}
+			}
+		};
+
+		const handleCancel = () => {
+			// Reset the input's local state to avoid
+			// stale values.
+			setInputValue( blockTitle );
+
+			onCancel();
+
+			// Must be assertive to immediately announce change.
+			speak( 'Leaving block name edit mode', 'assertive' );
+		};
+
+		const handleSubmit = () => {
+			// Submit changes only for ENTER.
+			onSubmit( inputValue );
+
+			let successAnnouncement;
+
+			if ( inputValue === '' ) {
+				successAnnouncement = __( 'Block name reset.' );
+			} else {
+				successAnnouncement = sprintf(
+					/* translators: %s: new name/label for the block */
+					__( 'Block name updated to: "%s".' ),
+					inputValue
+				);
+			}
+
+			// Must be assertive to immediately announce change.
+			speak( successAnnouncement, 'assertive' );
+		};
+
+		useEffect( () => {
+			speak( 'Entering block name edit mode', 'assertive' );
+		}, [] );
+
+		return (
+			<Popover
+				anchorRef={ ref }
+				placement={ 'overlay' }
+				resize={ true }
+				variant={ 'unstyled' }
+				animate={ false }
+				className="block-editor-list-view-block-node__rename-popover"
+			>
+				<form
+					className="block-editor-list-view-block-node__input"
+					onSubmit={ ( e ) => {
+						e.preventDefault();
+
+						onSubmit( inputValue );
+					} }
+					aria-describedby={ formDescriptionId }
+				>
+					<VisuallyHidden>
+						<p id={ formDescriptionId }>
+							{ __( 'Choose a custom name for this block.' ) }
+						</p>
+					</VisuallyHidden>
+					<InputControl
+						ref={ inputRef }
+						value={ inputValue }
+						label={ __( 'Edit block name' ) }
+						hideLabelFromVision
+						onChange={ ( nextValue ) => {
+							setInputValue( nextValue ?? '' );
+						} }
+						onBlur={ () => {
+							// Cancel editing mode.
+							handleCancel();
+						} }
+						onKeyDown={ onKeyDownHandler }
+						aria-describedby={ inputDescriptionId }
+						required
+					/>
+					<VisuallyHidden>
+						<p id={ inputDescriptionId }>
+							{ __(
+								'Press the ENTER key to submit or the ESCAPE key to cancel.'
```

Comment:
We could explore adding a visual submit button within the editing UI (perhaps similar to the one used in List View when you are _creating_ a link) which would ensure that it's absolutely clear how to submit the change for both visual and non-visual users.

Question: Would that also necessitate a cancel button or could we make that hidden given that `onBlur` will also cancel the editing.

Reply:
As far as I can tell, there is no way to reach this state with a screen reader/keyboard only. A visually hidden description intended to be used with screen readers while not allowing this state to be accessed via a screen reader makes me feel like we're not truly addressing the accessibility need here.

Reply:
@jeryj This is covered by `aria-describedby` on the input, line 131.

https://github.com/WordPress/gutenberg/pull/53852/files#diff-bf90c878e3f648941c5601515595134a3203fd7b98e1b5ab30fbe072b9b80a4cR131